go/arrow/compare.go

```diff
@@ -129,3 +129,15 @@ func TypeEqual(left, right DataType, opts ...TypeEqualOption) bool {
 		return reflect.DeepEqual(left, right)
 	}
 }
+
+// StorageTypeEqual checks if the underlying storage of two DataType are the same,
+// optionally checking metadata equality for STRUCT types by passing to TypeEqual
+func StorageTypeEqual(left, right DataType, opts ...TypeEqualOption) bool {
+	if leftExt, ok := left.(ExtensionType); ok {
+		left = leftExt.StorageType()
+	}
+	if rightExt, ok := right.(ExtensionType); ok {
+		right = rightExt.StorageType()
+	}
+	return TypeEqual(left, right, opts...)
+}
```

Comment:
might make sense to simply implement this as the default implementation of `ExtensionEquals` in `ExtensionBase` that other extension types can implement or otherwise. Honestly I don't see much benefit to having this function over letting consumers just call `TypeEqual` themselves and having the extension types implement `ExtensionEquals` appropriately.

Reply:
This is different than `ExtensionEquals`, as that one compares the underlying storage types and doesn't require both types to be extensions. In our case one of the two inputs is `FixedSizeBinaryType` and the is `UUIDType` which is an extension type. It's now only used in `UUIDType.deserialize()` so I've unexported and moved it to `internal/types`.